geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/FileSystemThreadedReader.scala

```diff
@@ -8,101 +8,327 @@
 
 package org.locationtech.geomesa.fs.storage.common
 
-import java.util.concurrent.{Executors, LinkedBlockingQueue, TimeUnit}
+import java.util.concurrent._
 
 import com.typesafe.scalalogging.StrictLogging
 import org.apache.hadoop.fs.Path
 import org.locationtech.geomesa.features.ScalaSimpleFeature
+import org.locationtech.geomesa.fs.storage.api.StorageMetadata.{StorageFileAction, StorageFilePath}
 import org.locationtech.geomesa.fs.storage.common.AbstractFileSystemStorage.FileSystemPathReader
 import org.locationtech.geomesa.utils.collection.CloseableIterator
 import org.locationtech.geomesa.utils.io.WithClose
 import org.opengis.feature.simple.SimpleFeature
 
 import scala.util.control.NonFatal
 
-object FileSystemThreadedReader {
+/**
+  * Multi-threaded file reads
+  *
+  * @param es executor service used by the read tasks
+  * @param phaser phaser for tracking read task completion
+  * @param queue intermediate queue populated by the read tasks
+  */
+class FileSystemThreadedReader private (es: ExecutorService, phaser: Phaser, queue: BlockingQueue[SimpleFeature])
+    extends CloseableIterator[SimpleFeature] {
 
-  def apply(
-      readers: Iterator[(FileSystemPathReader, Iterator[Path])],
-      threads: Int): CloseableIterator[SimpleFeature] = {
-    if (threads < 2) {
-      CloseableIterator(readers).flatMap { case (factory, paths) => CloseableIterator(paths).flatMap(factory.read) }
+  private val localQueue = new java.util.LinkedList[SimpleFeature]()
+
+  private var current: SimpleFeature = _
+
+  override def hasNext: Boolean = {
+    if (current != null) {
+      return true
+    }
+    current = localQueue.pollFirst
+    if (current != null) {
+      return true
+    }
+
+    while (!phaser.isTerminated) {
+      current = queue.poll(100, TimeUnit.MILLISECONDS)
+      if (current != null) {
+        queue.drainTo(localQueue, 10000)
+        return true
+      }
+    }
+    // last check - if phaser.isTerminated, the queue should have whatever values are left
+    current = queue.poll()
+    if (current != null) {
+      queue.drainTo(localQueue, 10000)
```

Comment:
If there are more than 10K objects in the queue, do they get dropped?  Seems like this may need to loop in order to drain a large queue...

Reply:
It doesn't need to drain the whole queue here - it will check it again on a subsequent call to `hasNext` once the local queue has been emptied.